packages/flutter_tools/lib/src/localizations/gen_l10n_types.dart

```diff
@@ -309,48 +313,96 @@ class Placeholder {
+          if (placeholder == null) {
+            throw L10nParserException(
+              'Make sure that the specified plural placeholder is defined in your arb file.',
+              filenames[locale]!,
+              resourceId,
+              messages[locale]!,
+              node.children[1].positionInMessage
+            );
+          }
+          placeholders[node.children[1].value!]!.isPlural = true;
+        }
+        if (node.type == ST.selectExpr) {
+          final Placeholder? placeholder = placeholders[node.children[1].value!];
+          if (placeholder == null) {
+            throw L10nParserException(
+              'Make sure that the specified select placeholder is defined in your arb file.',
+              filenames[locale]!,
+              resourceId,
+              messages[locale]!,
+              node.children[1].positionInMessage
+            );
+          }
+          placeholders[node.children[1].value!]!.isSelect = true;
+        }
+        traversalStack.addAll(node.children);
+      }
+    }
+    for (final Placeholder placeholder in placeholders.values) {
+      if (placeholder.isPlural && placeholder.isSelect) {
+        throw L10nException('Placeholder is used as both a plural and select in certain languages.');
+      } else if (placeholder.isPlural) {
+        if (placeholder.type == null) {
           placeholder.type = 'num';
         }
+        else if (!<String>['num', 'int'].contains(placeholder.type)) {
+          throw L10nException("Placeholders used in plurals must be of type 'num' or 'int'");
+        }
+      } else if (placeholder.isSelect) {
+        if (placeholder.type == null) {
+          placeholder.type = 'String';
+        } else if (placeholder.type != 'String') {
+          throw L10nException("Placeholders used in selects must be of type 'String'");
+        }
       }
-
-  static final RegExp _pluralRE = RegExp(r'\s*\{([\w\s,]*),\s*plural\s*,');
-  static final RegExp _selectRE = RegExp(r'\s*\{([\w\s,]*),\s*select\s*,');
+      placeholder.type ??= 'Object';
+    }
+  }
 
   final String resourceId;
   final String value;
   final String? description;
-  final List<Placeholder> placeholders;
-  final RegExpMatch? _pluralMatch;
-  final RegExpMatch? _selectMatch;
-
-  bool get isPlural => _pluralMatch != null && _pluralMatch!.groupCount == 1;
-  bool get isSelect => _selectMatch != null && _selectMatch!.groupCount == 1;
-
-  bool get placeholdersRequireFormatting => placeholders.any((Placeholder p) => p.requiresFormatting);
-
-  Placeholder getCountPlaceholder() {
-    assert(isPlural);
-    final String countPlaceholderName = _pluralMatch![1]!;
-    return placeholders.firstWhere(
-      (Placeholder p) => p.name == countPlaceholderName,
-      orElse: () {
-        throw L10nException('Cannot find the $countPlaceholderName placeholder in plural message "$resourceId".');
-      }
-    );
-  }
+  late final Map<LocaleInfo, String?> messages;
```

Comment:
If you thought it would make the rest of the implementation clearer, you could add some API here. Like:
```
message.translationFor(locale)
message.placeholdFor(name),
message.locales
etc..
```

Reply:
I'm not sure if I'll be able to set them this way since I run some code to also set `parsedMessages` and `filenames`.